### PR TITLE
cilium, ipvlan: avoid tail call in fast path

### PR DESCRIPTION
**Pull request by borkmann**
(Mainly in context of ipvlan to reduce latency.)

This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/6748)

### CODE REVIEW

**Comment by borkmann** [comment]
test-me-please

File: bpf/lib/tailcall.h [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+	__eval(__declare_tailcall_if_, COND)(NAME)
+
+#define __invoke_tailcall_if_0(NAME, FUNC)    \
+	return FUNC(skb)
+#define __invoke_tailcall_if_1(NAME, FUNC)    \
+	do {                                  \
+		ep_tail_call(skb, NAME);      \
+		ret = DROP_MISSED_TAIL_CALL;  \
+	} while (0)
+#define invoke_tailcall_if(COND, NAME, FUNC)  \
```
Comment (joestringer) [inline bpf/lib/tailcall.h:97]: I guess this is "invoke, or tail call if the condition COND is met" - it's admittedly getting longer but it'd be nice to make this clearer by either naming it `invoke_or_tailcall_if(COND...`, and/or adding a comment above this definition to describe it in more detail.

**Review by joestringer — APPROVED** [review]
I guess we could apply the same logic on ingress too? It was split in commit 7a98029b6b2cbc3a026d48ab2cf3637e53a7d985 but we were always enabling IPv6 at that time too.

Approving assuming that this doesn't break complexity on 4.9 kernels.

File: bpf/lib/tailcall.h (hunk 1 above)
Comment (borkmann) [reply bpf/lib/tailcall.h:97]: Will add a comment and integrate into policy bits too.

**Comment by borkmann** [comment]
test-me-please

**Comment by borkmann** [comment]
test-me-please

**Comment by borkmann** [comment]
test-me-please

**Comment by borkmann** [comment]
test-me-please

**Comment by borkmann** [comment]
test-me-please

**Comment by borkmann** [comment]
test-me-please

**Comment by borkmann** [comment]
test-me-please

**Review by tgraf — APPROVED** [review]

**Review by brb — APPROVED** [review]

**Comment by vadorovsky** [comment]
test-me-please

**Review by vadorovsky — APPROVED** [review]

**Review by ianvernon — APPROVED** [review]

**Comment by borkmann** [comment]
test-me-please

**Comment by borkmann** [comment]
test-me-please

**Pull request merged**